### PR TITLE
feat(nfts): kbw party ticket smart contracts

### PR DESCRIPTION
**Pull request by bearni95**
- [x] Basic smart contract implementation
- [x] Test coverage
- [x] Deployment scripts
- [x] NatSpec comments

## Contract 
- The contract deployer, `admin`, orchestrates the sale. 
  - The `payoutAddress` has no admin privileges
- The contract is `unpaused`, allowing any user to call `mint`
- When the winners are decided, `setWinners` is called by `admin`:
  - Contract becomes `paused` (cannot mint anymore)
  - Winners see their NFTs metadata updated automatically
  - Method `isWinner` allows checking for `tokenId` and `address` individual lookup
- When paused, `admin` can call `revokeAndReplaceWinner` to overwrite the win status of a ticket id for another
- The method `withdraw`, only callable by `admin` and when `paused`, transfers all the funds to `payoutAddress`
- `mintWinner` method mints a token to the specified `to` address, and sets it as a raffle winner. Must be called as an `admin`, and doesn't depend on the contract's pause system.

### CODE REVIEW

File: packages/nfts/contracts/party-ticket/TaikoPartyTicket.sol [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+    /// @param _blacklist The new blacklist address
+    function updateBlacklist(IMinimalBlacklist _blacklist) external onlyRole(DEFAULT_ADMIN_ROLE) {
+        blacklist = _blacklist;
+        emit BlacklistUpdated(address(_blacklist));
+    }
+
+    /// @notice Get individual token's URI
+    /// @param tokenId The token ID
+    /// @return The token URI
+    function tokenURI(uint256 tokenId) public view virtual override returns (string memory) {
```
Comment (2manslkh) [inline packages/nfts/contracts/party-ticket/TaikoPartyTicket.sol:87]:
```suggestion
    function tokenURI(uint256 tokenId) public view virtual override returns (string memory) {
        if (winners[tokenId]) {
            return string(abi.encodePacked(winnerBaseURI);
        }
        return string(abi.encodePacked(baseURI);
    }
```

File: packages/nfts/contracts/party-ticket/TaikoPartyTicket.sol [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+    /// @notice Winner base URI required to interact with IPFS
+    string public winnerBaseURI;
+    /// @notice Payout address
+    address public payoutAddress;
+    /// @notice Internal counter for token IDs
+    uint256 private _nextTokenId;
+    /// @notice Blackist address
+    IMinimalBlacklist public blacklist;
+    /// @notice Gap for upgrade safety
+    uint256[47] private __gap;
```
Comment (dantaik) [inline packages/nfts/contracts/party-ticket/TaikoPartyTicket.sol:46]:
```suggestion
    uint256[42] private __gap;
```

File: packages/nfts/contracts/party-ticket/TaikoPartyTicket.sol [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+        if (blacklist.isBlacklisted(to)) revert ADDRESS_BLACKLISTED();
+        uint256 tokenId = _nextTokenId++;
+        _safeMint(to, tokenId);
+    }
+
+    /// @notice Mint a winner ticket
+    /// @param to The address to mint to
+    /// @dev Requires calling as an admin
+    function mintWinner(address to) public onlyRole(DEFAULT_ADMIN_ROLE) {
+        if (blacklist.isBlacklisted(to)) revert ADDRESS_BLACKLISTED();
```
Comment (dantaik) [inline packages/nfts/contracts/party-ticket/TaikoPartyTicket.sol:177]: you can create a modifier for

```
if (blacklist.isBlacklisted(to)) revert ADDRESS_BLACKLISTED();
```

File: packages/nfts/contracts/party-ticket/TaikoPartyTicket.sol [hunk 4]
```diff
@@ -0,0 +1,229 @@
...
+    /// @param newWinnerId The ID of the new winner
+    function revokeAndReplaceWinner(
+        uint256 revokeId,
+        uint256 newWinnerId
+    )
+        external
+        whenPaused
+        onlyRole(DEFAULT_ADMIN_ROLE)
+    {
+        if (!winners[newWinnerId]) revert CANNOT_REVOKE_NON_WINNER();
```
Comment (dantaik) [inline packages/nfts/contracts/party-ticket/TaikoPartyTicket.sol:194]: how about create a `revokeWinners`, which can work with `setWinners` to replace `revokeAndReplaceWinner`?

File: packages/nfts/contracts/party-ticket/TaikoPartyTicket.sol [hunk 5]
```diff
@@ -0,0 +1,229 @@
...
+    /// @notice Unpause the contract
+    /// @dev Can only be called by the admin
+    function unpause() public onlyRole(DEFAULT_ADMIN_ROLE) {
+        _unpause();
+    }
+
+    /// @notice Withdraw the contract balance
+    /// @dev Can only be called by the admin
+    /// @dev Requires the contract to be paused
+    function withdraw() external whenPaused onlyRole(DEFAULT_ADMIN_ROLE) {
```
Comment (dantaik) [inline packages/nfts/contracts/party-ticket/TaikoPartyTicket.sol:214]: need `whenPaused`?

**Review by dantaik — CHANGES_REQUESTED** [review]

File: packages/nfts/test/party-ticket/TaikoPartyTicket.t.sol [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+pragma solidity 0.8.24;
+
+import { Test } from "forge-std/src/Test.sol";
+
+import { TaikoPartyTicket } from "../../contracts/party-ticket/TaikoPartyTicket.sol";
+import { Merkle } from "murky/Merkle.sol";
+import { ERC1967Proxy } from "@openzeppelin/contracts/proxy/ERC1967/ERC1967Proxy.sol";
+import { MockBlacklist } from "../util/Blacklist.sol";
+
+contract TaikoPartyTicketTest is Test {
```
Comment (dantaik) [inline packages/nfts/test/party-ticket/TaikoPartyTicket.t.sol:11]: can you test upgrading the implementation, not sure if this is supported by your code.

File: packages/nfts/contracts/party-ticket/TaikoPartyTicket.sol [hunk 7]
```diff
@@ -0,0 +1,229 @@
...
+import { AccessControlUpgradeable } from
+    "@openzeppelin/contracts-upgradeable/access/AccessControlUpgradeable.sol";
+import { PausableUpgradeable } from
+    "@openzeppelin/contracts-upgradeable/utils/PausableUpgradeable.sol";
+import { IMinimalBlacklist } from "@taiko/blacklist/IMinimalBlacklist.sol";
+
+/// @title TaikoPartyTicket
+/// @dev ERC-721 KBW Raffle & Party Tickets
+/// @custom:security-contact security@taiko.xyz
+contract TaikoPartyTicket is
```
Comment (dantaik) [inline packages/nfts/contracts/party-ticket/TaikoPartyTicket.sol:20]: is this contract upgradeable through a proxy?

File: packages/nfts/contracts/party-ticket/TaikoPartyTicket.sol (hunk 4 above)
Comment (bearni95) [reply packages/nfts/contracts/party-ticket/TaikoPartyTicket.sol:202]: Adding the `revokeWinner` method. I'd leave in the `revokeAndReplaceWinner` to ease hot-fixing the winners list.

File: packages/nfts/test/party-ticket/TaikoPartyTicket.t.sol [hunk 8]
```diff
@@ -0,0 +1,171 @@
...
+        // run winner flow
+        test_winnerFlow();
+        // ensure URIs are "winner" and "loser" after setting winners
+        assertEq(token.tokenURI(0), "ipfs://baseURI/winner.json");
+        assertEq(token.tokenURI(1), "ipfs://baseURI/loser.json");
+    }
+
+    function test_upgrade() public {
+        // create the v1 instance of the token
+        vm.startBroadcast(admin);
```
Comment (dantaik) [inline packages/nfts/test/party-ticket/TaikoPartyTicket.t.sol:127]: please avoid using start/stopBroadcast in tests, use start/stopPrank instead.

File: packages/nfts/test/party-ticket/TaikoPartyTicket.t.sol [hunk 9]
```diff
@@ -0,0 +1,171 @@
...
+        tokenV1.mint{ value: MINT_FEE }();
+        vm.prank(minters[1]);
+        tokenV1.mint{ value: MINT_FEE }();
+        vm.prank(minters[2]);
+        tokenV1.mint{ value: MINT_FEE }();
+
+        // upgrade to v2
+        vm.startBroadcast(admin);
+
+        tokenV1.upgradeToAndCall(
```
Comment (dantaik) [inline packages/nfts/test/party-ticket/TaikoPartyTicket.t.sol:154]: try to prank a non-admin, to verify the upgradeToAndCall revert, then prank the admin to see the same tx succeed

**Review by dantaik — CHANGES_REQUESTED** [review]

**Review by dantaik — APPROVED** [review]

**Review by KorbinianK — APPROVED** [review]

**Pull request merged**